Use defaults from Schema as base for Data Values

pivotaljohn opened the pull request.

gcheadle-vmware committed:
Use defaults from Schema as base for Data Values

- set default array values in schemas to be empty
- when schema feature is enabled, but no schema provided represent as `NullSchema`.
- when no schema given, but data values are specified, error.
- group logic invoked when schema flag enabled

Signed-off-by: Garrett Cheadle <gcheadle@vmware.com>
Signed-off-by: John Ryan <jtigger@infosysengr.com>
Signed-off-by: Cari Dean <cdean@pivotal.io>

pivotaljohn commented:
@cppforlife — if you get a moment, this is a good spot for review/feedback of all kinds.

StevenLocke commented on pkg/cmd/template/schema_test.go:
```diff
@@ -13,8 +13,115 @@ import (
+		dataValuesYAML := `#@data/values
+---
+db_conn:
+#@overlay/append
```
Nice to have:
This feels off to me. We're not really appending anything, we need an "overlay/match" annotation and append gives us one for free. This likely already existed before this commit :)

pivotaljohn replied:
Yes. Good news is that this is not a degradation in UX: before schemas, Configuration Authors have supplied a base Data Values that illustrated that a particular Data Value is a list. Configuration Consumers have already had to supply some kind of `@overlay` annotation on their arrays.

Even better news is that in the very next story, we're aiming to improve that UX further by setting a default behavior (likely `@overlay/replace` on the Array itself) so that, Configuration Consumers can provide configuration _without_ having to learn about overlays, first. 👍🏼   Stay tuned!

StevenLocke commented on pkg/cmd/template/schema_test.go:
```diff
@@ -13,8 +13,115 @@ import (
+`
+		dataValuesYAML := `#@data/values
+---
+#@overlay/append
+- first
+#@overlay/append
+- second
```
Same nice to have, this doesn't seem like an append is what we're aiming for, just trying to match.

pivotaljohn replied:
... and to be clear, the `@overlay/append`s _are_ needed here... one needs to provide _both_ an `@overlay/match` _and_ an operation (e.g. `@overlay/replace`)... it's just that `@overlay/append` is semantically equivalent to `overlay/match` and `@overlay/insert after`.

StevenLocke commented on pkg/cmd/template/schema_test.go:
```diff
@@ -24,666 +131,523 @@ vpc:
 
-func TestMapOnlySchemaChecksOk(t *testing.T) {
-	schemaYAML := `#@schema/match data_values=True
+func TestDataValueNotConformingToSchemaFails(t *testing.T) {
```
Nice to have:
It would be great if all of these "expected failure" tests could use a test helper like the success cases. Instead of performing the setup and checking out.Err, passing an expected error string may be sufficient.

DennisDenuto replied:
Thanks. we have moved the failure test cases to a test helper as suggested.

StevenLocke commented on pkg/cmd/template/schema_test.go:
```diff
@@ -24,666 +131,523 @@ vpc:
+---
+vpc:
+  subnet_config:
+  #@overlay/append
   - id: 2
+  #@overlay/append
   - id: 3
```
Same NTH, match > append?

DennisDenuto replied:
the @overlay/appends are needed here... one needs to provide both an @overlay/match and an operation (e.g. @overlay/replace)... it's just that @overlay/append is semantically equivalent to overlay/match and @overlay/insert after.

StevenLocke commented on pkg/cmd/template/schema_test.go:
```diff
@@ -24,666 +131,523 @@ vpc:
+}
 
-	ui := cmdcore.NewPlainUI(false)
+func TestSchemaFeatureIsNotEnabledButSchemaIsPresentReportsAWarning(t *testing.T) {
```
Are we missing an assertion here? This name seems to indicate that a warning should exist, but I don't see any verification for it.

pivotaljohn replied:
Yup, @DennisDenuto and I saw this too. We decided to defer this to a chore: https://www.pivotaltracker.com/story/show/176465929

danielhelfand replied:
@pivotal-dean and I added a NewFakeUI approach in this [commit](https://github.com/vmware-tanzu/carvel-ytt/commit/3429491437f2a42ed066f1cee3787eac8e51d136). Hopefully should help with these cases in future.

pivotaljohn replied:
Yes! Let's rebase and make use of that approach, here.

DennisDenuto replied:
Nice! we proceeded with the testing approach here in `schema_test.go`
We made some changes, the main one being that cmds now accept a `UI` interface. This allowed for the test code to implement a 'fakeUI' and pass that to a command. The upside being that we could remove the 'NewFakeUI' function that lived in the `template` package.

StevenLocke commented on pkg/cmd/template/schema_test.go:
```diff
@@ -711,57 +673,25 @@ rendered: true`
-	}
-}
-
 func testSchemaTemplates(t *testing.T, schemaYAML string, dataValuesYAML string, templateYAML string, expected string) {
```
This is an awesome test helper! It makes the tests so much cleaner.

As we've been reviewing the changes, we thought of a couple ways to augment it.
Accept an error message parameter to denote when the `RunWithFiles ` call should fail?
Switch the file string params to a `filesToProcess` parameter? This would allow not providing the dataValues or providing multiple.

DennisDenuto replied:
Good points.
filesToProcess is now a parameter that is passed to the helper function.

Also, instead of passing an error argument as suggested, we went with splitting the 'assertion' helper function into 2 separate functions:
1) assertYttWorkflowSucceedsWithOutput
2) assertYttWorkflowFailsWithErrorMessage

StevenLocke commented on pkg/cmd/template/cmd.go:
```diff
@@ -132,30 +131,32 @@ func (o *TemplateOptions) RunWithFiles(in TemplateInput, ui cmdcore.PlainUI) Tem
+	var libraryValues []*workspace.DataValues
+
+	if o.SchemaEnabled {
+		if len(schemaDocs) > 0 {
```
Will there ever be a case where we want to support multiple schema files? If not, should we simply error here instead of taking the first one? (This has been the case for a while, not directly related to this change set)

DennisDenuto replied:
Yes we will eventually want to support multiple schemas being provided to ytt and is being prioritized and hopefully worked on soon.

For now, this experimental schema feature will quietly process the first schema file with the intent that multiple schema files will be allowed and processed in the future.

StevenLocke commented on pkg/workspace/data_values_pre_processing.go:
```diff
@@ -39,6 +39,7 @@ func (o DataValuesPreProcessing) Apply() (*DataValues, []*DataValues, error) {
 
 func (o DataValuesPreProcessing) apply(files []*FileInLibrary) (*DataValues, []*DataValues, error) {
 	var values *yamlmeta.Document
+	values = o.loader.schema.AsDataValue()
```
Nit: Any reason not to use `:=`?

DennisDenuto replied:
👍

StevenLocke commented on pkg/schema/schema.go:
```diff
@@ -5,23 +5,43 @@ package schema
-	Name    string
-	Source  *yamlmeta.Document
-	Allowed *DocumentType
+	Name       string
+	Source     *yamlmeta.Document
```
These don't seem to be read anywhere. Is there a future requirement we're setting ourselves up for?

DennisDenuto replied:
There is a near term future requirement where we want to report error messages relating to schema type check. And for that to work we will need to have a reference to the schema source and the name field.

StevenLocke commented on pkg/schema/schema.go:
```diff
@@ -5,23 +5,43 @@ package schema
 type AnySchema struct {
 }
 
+type NullSchema struct {
```
As far as I can tell, the only distinction between AnySchema and NullSchema is in `ValidateWithValues`. Are there other benefits to creating this struct?

DennisDenuto replied:
Yes that is the only distinction between AnySchema and NullSchema. My understanding is that we are using polymorphism here to communicate different behavior for when a schema isn't provided vs when a schema is completely permissive. John has added this as a tech forum topic to discuss more as a team.

StevenLocke reviewed (commented):
In the [glossary](https://hackmd.io/pODV3wzbT56MbQTxbQOOKQ#Glossary) we define "type checking" but our error messages call it "typechecking". Not directly related to this change set, but wanted to note it somewhere.

joaopapereira commented on pkg/schema/schema.go:
```diff
@@ -143,10 +158,63 @@ func newCollectionItemValueType(collectionItemValue interface{}) (yamlmeta.Type,
 	return nil, fmt.Errorf("Collection item type did not match any known types")
 }
 
+func DefaultDataValues(doc *yamlmeta.Document) (*yamlmeta.Document, error) {
```
What is the intent behind make this a public method? Is the idea that we will be able to use this externally?

Also, this feels like a factory, should it be named `NewDefaultDataValues`?

DennisDenuto replied:
Thanks, we have made this function private. This isn't a 'factory' function, it doesn't construct different data value structs, instead it modifies the defaults from the schema document to produce the base for data values.

joaopapereira commented on pkg/schema/schema.go:
```diff
@@ -5,23 +5,43 @@ package schema
 
 func NewDocumentSchema(doc *yamlmeta.Document) (*DocumentSchema, error) {
-	docType := &DocumentType{Source: doc}
+	docType, err := NewDocumentType(doc)
```
An idea:
What if DocumentSchema had a #Parse() error that would internally calculate both the docType and defaultDVs?
This would prevent us from have dangling functions

DennisDenuto replied:
It is a good observation. We think that the current implementation is sufficient for the moment. We should discuss this further. We are going to proceed with this PR as-is. 

Let's continue with your idea further on this github comment thread. And let's see if we can incorporate this idea as part of a future story/chore.

joaopapereira commented on pkg/schema/schema.go:
```diff
@@ -143,10 +158,63 @@ func newCollectionItemValueType(collectionItemValue interface{}) (yamlmeta.Type,
+	return docCopy.(*yamlmeta.Document), nil
+}
+
+func setDefaultValues(node interface{}) {
```
Just some food for thought
Having generic `interface{}` in a couple of these methods feels off. Personally, I would like to have the compiler help to ensure we are using the correct types here.
Maybe having an interface called `yamlmeta.Element` or `yamlmeta.Node` or `yamlmeta.Item` not sure naming is hard.

DennisDenuto replied:
Thanks, we are now using the `yamlmeta.Node` type instead of an empty interface. And we are also guarding against values that are not nodes (i.e. scalars)

joaopapereira commented on pkg/schema/schema.go:
```diff
@@ -143,10 +158,63 @@ func newCollectionItemValueType(collectionItemValue interface{}) (yamlmeta.Type,
+
+func (n NullSchema) ValidateWithValues(valuesFilesCount int) error {
+	if valuesFilesCount > 0 {
+		return fmt.Errorf("Schema experiment flag was enabled but no schema document was provided")
```
This message feels like it is at the wrong level of abstraction. Should the Schema package be aware that there is a flag in cmd that is experimental?

DennisDenuto replied:
We are going to drop reference to the 'flag' word. And make the error message a bit more generic. "Schema feature is enabled but no schema document was provided"

joaopapereira commented on pkg/schema/schema.go:
```diff
@@ -143,10 +158,63 @@ func newCollectionItemValueType(collectionItemValue interface{}) (yamlmeta.Type,
+	return nil
+}
+
+func (n NullSchema) ValidateWithValues(valuesFilesCount int) error {
```
I am a little bit confused about this function. Is the intent of the function to error out if a schema is NOT provided and no data values files are present?
Not sure if Schema should be responsible for this check. Maybe the LibraryLoader should have `#HasSchemaFiles() bool` and `#HasValuesFiles() bool`. This way at the `cmd` level we can check if the above is conditions are met to continue or not.

DennisDenuto replied:
> I am a little bit confused about this function. Is the intent of the function to error out if a schema is NOT provided and no data values files are present?

No, the intent is for when schema is enabled, no schema is provided, and there are data values. Then it should error out.

It is a good observation. We think that the current implementation is sufficient for the moment. We should discuss this further. We've added it to tech forum.

joaopapereira reviewed (commented):
Some comments around the structure.

pivotaljohn commented:
> In the [glossary](https://hackmd.io/pODV3wzbT56MbQTxbQOOKQ#Glossary) we define "type checking" but our error messages call it "typechecking". Not directly related to this change set, but wanted to note it somewhere.

Thanks, @StevenLocke — noted in the upcoming story about taking a thorough charge at error messaging. 👍🏼

DennisDenuto committed:
Reorganize Schema integration tests

... from a flat set of 20+ tests into groups by context and/or feature.

Co-authored-by: John Ryan <jtigger@infosysengr.com>

DennisDenuto committed:
Make it easier for a test to make an assertion when a cmd writes to the 'ui'

- cmds now accept a UI interface instead of a concrete type (PlainUI)
- Add a test implementation 'fakeui' of the UI interface to make assertions when
a cmd writes to the UI
- Move UI interface to the 'ui' package
- Move core/plainUI -> ui/tty

Co-authored-by: Dennis Leon <leonde@vmware.com>
Co-authored-by: John Ryan <jtigger@infosysengr.com>

DennisDenuto committed:
Code refactor

- stop exposing a helper function (defaultDataValues) in the schema package
- use a type 'yamlmeta.Node' instead of an empty interface when setting
default values

Co-authored-by: Dennis Leon <leonde@vmware.com>
Co-authored-by: John Ryan <jtigger@infosysengr.com>

danielhelfand commented on pkg/cmd/template/regular_input.go:
```diff
@@ -5,9 +5,9 @@ package template
 
 import (
 	"fmt"
+	ui2 "github.com/k14s/ytt/pkg/cmd/ui"
```
Maybe it would make sense to not alias `github.com/k14s/ytt/pkg/cmd/ui`? It has two different aliases now (e.g. `ui2` and `cmdui`, so it would be nice to pick a single alias across the whole project.

danielhelfand commented on pkg/cmd/template/regular_input_test.go:
```diff
@@ -155,3 +164,24 @@ func assertStdoutAndStderr(stdout *bytes.Buffer, stderr *bytes.Buffer, expectedS
 	return nil
 }
+
+type fakeUI struct {
+	stdout io.Writer
+	stderr io.Writer
+}
+
+func (ui fakeUI) DebugWriter() io.Writer {
+	return ui.stderr
+}
+
+func (ui fakeUI) Printf(str string, args ...interface{}) {
+	fmt.Fprintf(ui.stdout, str, args...)
+}
+
+func (ui fakeUI) Warnf(str string, args ...interface{}) {
+	fmt.Fprintf(ui.stderr, str, args...)
+}
+
+func (ui fakeUI) Debugf(str string, args ...interface{}) {
+	fmt.Fprintf(ui.stderr, str, args...)
+}
```
Moving this into this block of tests makes it harder to find this pattern for asserting on ui output streams in my opinion. I think it would make sense to move under the ui package so it can be more easily found/used.

I still do think there should be a constructor for it or its fields should be public since tests are in a separate package.
